File: app/cron/backup.rb
require 'open3'
require 'shellwords'

def cron_backup_res
  return unless ENV['DB_USERNAME']

  current_file_folder = File.dirname(__FILE__)
  src = "#{current_file_folder}/../res"
  dst = "#{current_file_folder}/../res.swp"

  escaped_username = Shellwords.escape(ENV['DB_USERNAME'])
  escaped_password = Shellwords.escape(ENV['DB_PASSWORD'])

  dump_command = "mariadb-dump -u #{escaped_username} --password=#{escaped_password} --databases taskcentral | gzip -9 > res/mariadb_dump.sql"

  rsync_exists = system("command -v rsync >/dev/null 2>&1")

  if rsync_exists
    command = "rsync -a #{src}/ #{dst}/"
  else
    command = "cp -r --preserve=all #{src}/ #{dst}/"
  end

  puts Time.now
  puts "CRON JOB: Running #{dump_command}"

  system(dump_command)

  puts Time.now
  puts "CRON JOB: Running #{command}"

  stdout, stderr, status = Open3.capture3(command)

  if status.success?
    puts "Copy successful."
  else
    puts "Error: #{stderr}"
  end
end

# Backup res to res.swp every 2 AM PDT (4 AM CDT)
$CRON_SCHEDULER.cron "0 2 * * *" do
  cron_backup_res
end

cron_backup_res

File: src/cron/backup.rs


use std::process::{Command, Stdio};
use std::env;
use std::io;
use std::fs::File;
use std::io::Write;

use crate::controllers::app_controller::SharedState;
use crate::config::AppContext;

pub fn cron_backup_res(state: &SharedState) -> io::Result<()> {
    if let Some(db_username) = env::var("DB_USERNAME").ok() {
        let acontext: &AppContext = &*state.clone();

        let src = format!("{}/res", &acontext.app_root);
        let dst = format!("{}/res.swp", &acontext.app_root);

        println!("{} CRON JOB: Running mariadb-dump", chrono::Local::now());

        let mariadb_dump = Command::new("mariadb-dump")
            .arg("-u")
            .arg(&db_username)
            .arg("--password=".to_string() + &env::var("DB_PASSWORD").unwrap())
            .arg("--databases")
            .arg("taskcentral")
            .stdout(Stdio::piped())
            .spawn()?;
        let gzip = Command::new("gzip")
            .arg("-9")
            .stdin(mariadb_dump.stdout.ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Failed to open mariadb-dump stdout"))?)
            .stdout(Stdio::piped())
            .spawn()?;

        let output_file = format!("{}/mariadb_dump.sql.gz", src);
        let mut file = File::create(output_file)?;
        let gzip_output = gzip.wait_with_output()?;
        file.write_all(&gzip_output.stdout)?;

        println!("{} CRON JOB: mariadb-dump to gzip completed.", chrono::Local::now());


        let rsync_exists = Command::new("sh").arg("-c")
            .arg("command -v rsync >/dev/null 2>&1").status()?;
        let command = if rsync_exists.success() {
            format!("rsync -a {}/ {}/", src, dst)
        } else {
            format!("cp -r --preserve=all {}/ {}/", src, dst)
        };

        println!("{} CRON JOB: Running {}", chrono::Local::now(), command);

        let output = Command::new("sh").arg("-c").arg(&command).output()?;

        if output.status.success() {
            println!("Copy successful.");
        } else {
            eprintln!("Error: {}", String::from_utf8_lossy(&output.stderr));
        }
    }
    Ok(())
}
